odoo/models.py

```diff
@@ -3469,37 +3474,111 @@ def _check_parent_path(self):
             _logger.error('parent_path field on model %r should be indexed! Add index=True to the field definition.', self._name)
 
     def _add_sql_constraints(self):
-        """ Modify this model's database table constraints so they match the one
-        in _sql_constraints.
+        """ Modify this model's database table constraints and indexes
+        so they match the ones in _sql_objects.
+        """
+        for cons in self._sql_objects.values():
+            cons.sync_database_object(self)
 
+    @api.model
+    def __get_columns_from_sql_diagnostics(self, diagnostics, *, check_registry=False) -> list[str]:
+        """Given the diagnostics of an error, return the affected column names by the constraint.
+        Return an empty list if we cannot determine the columns.
         """
-        cr = self._cr
-        foreign_key_re = re.compile(r'\s*foreign\s+key\b.*', re.I)
-
-        for (key, definition, message) in self._sql_constraints:
-            conname = '%s_%s' % (self._table, key)
-            if len(conname) > 63:
-                hashed_conname = sql.make_identifier(conname)
-                current_definition = sql.constraint_definition(cr, self._table, hashed_conname)
-                if not current_definition:
-                    _logger.info("Constraint name %r has more than 63 characters, internal PG identifier is %r", conname, hashed_conname)
-                conname = hashed_conname
-            else:
-                current_definition = sql.constraint_definition(cr, self._table, conname)
-            if current_definition == definition:
-                continue
+        if column := diagnostics.column_name:
+            return [column]
+        if not check_registry:
+            return []
+        # start a new transaction because the current transaction is probably already failing
+        with self.pool.cursor() as cr_tmp:
```

Comment:
Should be manage by the upper method because you still need to have the translation